Make it possible to cache memory between invocations

koute opened the pull request:
I've added `Dict::tokenize` which is basically `parse_to_lexertokens` but allows you to 1) pass a `Cache` object which when used across multiple invocations will let the code reuse the previously made internal allocations, and 2) pushes the tokens to an existing vector instead of returning a new one. This makes it possible to run the whole thing without making any allocations in the average case.

This speeds up one of my benchmarks by 30% when there already exists pressure on the allocator (basically, when something else is already running in the background and hammering the allocator), and speeds it up by 20% if there isn't any allocator pressure (when only the benchmark code itself is running).

I've left `parse_to_lexertokens` for backwards compatibility, however I've slapped an `#[deprecated]` on it. I've left `parse` alone, though personally I'd just remove it altogether - after all it's trivial to just call `read_feature_string` yourself. (It's up to you though.)

koute committed:
Add `Dict::tokenize` and `Dict::tokenize_with_cache`

wareya commented:
There are a *lot* of use cases for tools like this where throughput just doesn't matter at all, like "parsers" (the reading-assistance tools). It makes more sense to me to provide three different methods of parsing with increasing levels of work on the part of the caller when this is the case, assuming that each is basically just wrapping the next more complex one with a small amount of boilerplate (which is the case here).

I'll take a look at the rest of this once my head is out of the toy-programming-language-design gutter.

koute commented:
I agree that there are a lot of use cases when speed doesn't matter all that much. However, in my opinion, in this case the boilerplate is so tiny compared to all of the rest of the surrounding code the user will have to write *anyway* (e.g. my code which is only parsing the feature strings into something usable is over 200 lines long, never mind everything else that actually tries to do something interesting with this data) that it might not be worth it maintaining all of those separate APIs. And with the most low-level variant the user can explicitly see that both the `Vec` and the `Cache` are being recreated every time, so they're aware of the inefficiency.

If you compare this:

```rust
let mut tokens = Vec::new();
let mut cache = Cache::new();
if let Ok(_) = dict.tokenize(&mut cache, text, &mut tokens) {
    for token in tokens {
        let slice = &text[token.start..token.end];
        let feature = dict.read_feature_string(&token);
    }
}
```

to this:

```rust
if let Some((tokens, _)) = parse(&dict, text) {
    for token in tokens {
        let slice = token.surface;
        let feature = token.feature;
    }
}
```

they're not *that* much different when in comes to ease of use. (Especially for an API which will be only called once or twice in the whole application.)

But, again, it's up to you and I don't really care either way. (: If you'd like to keep different levels of ergonomy/inefficiency tradeoff I'd suggest that we get rid of `parse_to_lexertokens` and `parse`, and either:

1) have something like `Dict::tokenize_with_cache`, `Dict::tokenize` and `Dict::parse` with increasing levels of ergonomy,
2) or have something like `Dict::tokenize_with_cache` and `Dict::tokenize`, and instead of `Dict::parse` and `ParserToken` just add `LexerToken::get_feature(&self, &Dict)` to get the feature string and convert the `LexerToken::{start, end}` fields into a `Range<usize>` so that you can just do `&whole_text[token.range]`.

wareya commented:
Yeah, those both seem like good ideas.

koute committed:
Replace `LexerToken::{start, end}` with `LexerToken.range`

koute committed:
Add `LexerToken::get_text` and `LexerToken::get_feature`

koute committed:
Remove `parse_to_lexertokens`

koute committed:
Remove `ParserToken` and `parse`

koute commented:
Okay, done.

I've added two methods - `tokenize` and `tokenize_with_cache`, removed the freestanding functions, got rid of `ParserToken` and added helper methods to `LexerToken` to get the feature string and the text chunk. I think this now makes the API simpler without introducing a hit to ergonomy.

This obviously is not backwards compatible.

wareya commented:
All good.
